### PR TITLE
feat: 188 create class for reinforcement order

### PR DESCRIPTION
**Pull request by ArdtK**
## Issue addressed
Solves #188 

## Code of conduct
- [x] I HAVE NOT added sensitive or compromised (test) data to the repository.
- [x] I HAVE NOT added vulnerabilities to the repository.
- [ ] I HAVE discussed my solution with (other) members of the KOSWAT team.

## What has been done?
- Add class `OrderStrategyReinforcements` for providing the reinforcement types to the order strategy. This class implements new protocol `StrategyReinforcementsProtocol` to prepare for other strategies in the future.
- Extend `StrategyReinforcementTypeCosts` with the profiles surface and rename it to `StrategyReinforcementType`.
- Extend `StrategyInput` with the reinforcements.

### Checklist
- [x] Tests are either added or updated.
- [x] Branch is up to date with `master`.
- [x] Updated documentation if needed.

## Additional Notes (optional)
- Fix some typehints

### CODE REVIEW

File: koswat/cost_report/summary/koswat_summary_builder.py [hunk 1]
```diff
@@ -94,13 +95,22 @@ def _get_final_reinforcement_per_location(
...
+    ) -> list[StrategyLocationReinforcement]:
         _matrix = KoswatSummaryLocationMatrixBuilder(
             available_locations=available_locations,
             locations_profile_report_list=locations_profile_report_list,
         ).build()
+        _reinforcements = [
+            StrategyReinforcementType(
+                reinforcement_type=type(x.profile_cost_report.reinforced_profile),
+                base_costs=x.profile_cost_report.total_cost,
+                ground_level_surface=x.profile_cost_report.reinforced_profile.new_ground_level_surface,
```
Comment (Carsopre) [inline koswat/cost_report/summary/koswat_summary_builder.py:107]: Open question, why would we do this? We are better just setting the `reinforcement` property, then you can via a property `reinforcement_type` return directly the type. This way we keep all available data available (at the expense of memory though).

Once I finish the review I might change my opinion though.

File: koswat/strategies/strategy_reinforcement_type.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from dataclasses import dataclass
+from typing import Type
+
+from koswat.dike_reinforcements.reinforcement_profile.reinforcement_profile_protocol import (
+    ReinforcementProfileProtocol,
+)
+
+
+@dataclass
+class StrategyReinforcementType:
```
Comment (Carsopre) [inline koswat/strategies/strategy_reinforcement_type.py:10]: Why would you change the name and the class location? The whole idea of this dataclass is to proxy costs per reinforcement type. Keep in mind doing these changes during parallel work brings conflicts to the other parties (me).
Revert.

File: koswat/cost_report/summary/koswat_summary_location_matrix_builder.py [hunk 3]
```diff
@@ -30,11 +28,11 @@ class KoswatSummaryLocationMatrixBuilder(BuilderProtocol):
 
     def _get_multi_location_profile_to_dict_matrix(
         self, locations_profile: MultiLocationProfileCostReport
-    ) -> dict[PointSurroundings, StrategyReinforcementTypeCosts]:
+    ) -> dict[PointSurroundings, StrategyReinforcementType]:
```
Comment (Carsopre) [inline koswat/cost_report/summary/koswat_summary_location_matrix_builder.py:31]: Don't change names of classes without previous agreement, specially when working in parallel branches.

File: koswat/strategies/README.md [hunk 4]
```diff
@@ -9,15 +9,16 @@ This modules contains the logic to choose which measure will be applied for a gi
 
 - `StrategyLocationInput`, gathers all the input required for a strategy to determine which reinforcement can be applied based on its location (`point_surrounding`) and `available_reinforcements`.
     - point_surrounding (`PointSurroundings`), a point (meter) in the dike traject.
-    - strategy_reinforcement_type_costs (`list[StrategyReinforcementTypeCosts]`), all the reinforcements that can be used at this location and their related costs.
-    - cheapest_reinforcement (`StrategyReinforcementTypeCosts`), returns which "available reinforcment" has the lower total costs at this location.
-    - available_measures (`Type[ReinforcementProfileProtocol]`), returns only the reinforcement type from the `strategy_reinforcement_type_costs` collection.
+    - strategy_reinforcement_type (`list[StrategyReinforcementType]`), all the reinforcements that can be used at this location and their related costs.
```
Comment (Carsopre) [inline koswat/strategies/README.md:12]: As said, don't do this without previous agreement during parallel work, unnecessary effort that brings undesired conflicts.

File: koswat/strategies/README.md [hunk 5]
```diff
@@ -9,15 +9,16 @@ This modules contains the logic to choose which measure will be applied for a gi
...
+    - cheapest_reinforcement (`StrategyReinforcementType`), returns which "available reinforcment" has the lower total costs at this location.
+    - available_measures (`Type[ReinforcementProfileProtocol]`), returns only the reinforcement type from the `strategy_reinforcement_type` collection.
 
-- `StrategyReinforcementTypeCosts`, maps a type of reinforcement to the calculated costs from the `cost_report` subproject.
+- `StrategyReinforcementType`, maps a type of reinforcement to the calculated costs from the `cost_report` subproject.
     - reinforcement_type (`Type[ReinforcementProfileProtocol]`), the mapped reinforcement type.
     - base_costs (`float`), the costs only related to the reinforcement's required space (thus excluding infrastructure costs).
     - infrastructure_costs (`float`), the costs associated **only** to infrastructures.
     - total_costs (`float`), the addition of `base_costs` and `infrastructure_costs`.
+    - ground_level_surface (`float`), the surface (width) of the ground level.
```
Comment (Carsopre) [inline koswat/strategies/README.md:21]: is this so? from toe to toe? (be explicit about this please). `CharacteristicPoints.p0` to `CharacteristicPoints.p8` ?

File: koswat/strategies/order_strategy/order_strategy.py [hunk 6]
```diff
@@ -72,7 +63,9 @@ def get_strategy_reinforcements(
     def apply_strategy(
         self, strategy_input: StrategyInput
     ) -> list[StrategyLocationReinforcement]:
-        _reinforcement_order = self.get_default_order_for_reinforcements()
+        _reinforcement_order = OrderStrategyReinforcements.from_strategy_input(
+            strategy_input
+        ).strategy_reinforcements
```
Comment (Carsopre) [inline koswat/strategies/order_strategy/order_strategy.py:68]: It does not make sense to me to use a class method for this when the only usecase for this class is delivering the `strategy_reinforcements`.

File: koswat/strategies/strategy_location_input.py
Comment (Carsopre) [inline koswat/strategies/strategy_location_input.py:1]: unrequired changes.

File: koswat/strategies/strategy_reinforcements_protocol.py
Comment (Carsopre) [inline koswat/strategies/strategy_reinforcements_protocol.py:1]: What is the purpose of this protocol? You are only using it once on a concrete instance. From what I see it needs to be removed.

File: koswat/strategies/order_strategy/order_strategy_reinforcements.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+)
+from koswat.strategies.strategy_input import StrategyInput
+from koswat.strategies.strategy_reinforcement_type import StrategyReinforcementType
+from koswat.strategies.strategy_reinforcements_protocol import (
+    StrategyReinforcementsProtocol,
+)
+
+
+@dataclass
+class OrderStrategyReinforcements(StrategyReinforcementsProtocol):
```
Comment (Carsopre) [inline koswat/strategies/order_strategy/order_strategy_reinforcements.py:31]: I do not think a protocol is required at all, at least not at this moment.

File: koswat/strategies/order_strategy/order_strategy_reinforcements.py
Comment (Carsopre) [inline koswat/strategies/order_strategy/order_strategy_reinforcements.py:1]: This class should not exist, it is simply overtaking the responsibility previously set in `OrderStrategy`

**Review by Carsopre — COMMENTED** [review]
Unfortunately too many things need to be changed from my side, I do not think it's worth finishing the review without rethinking the whole purpose of this PR   / issue.

**Comment by ArdtK** [comment]
Not merged as 2nd implementation was created #201

**Pull request closed without merging**